wrappers/services.go

```diff
@@ -75,30 +76,107 @@ func generateSnapServiceFile(app *snap.AppInfo, opts *AddSnapServicesOptions) ([
 	return genServiceFile(app, opts)
 }
 
+// max returns the maximum of two integers. Why is this
+// not provided by golang?
+func max(a, b int) int {
+	if a < b {
+		return b
+	}
+	return a
+}
+
+func min(a, b int) int {
+	if b < a {
+		return b
+	}
+	return a
+}
+
+func formatCpuGroupSlice(grp *quota.Group) string {
+	if grp.CpuLimit == nil {
+		return ""
+	}
+
+	// calculateSystemdCPULimit calculates the number of cpus and the allowed percentage
+	// to the systemd specific format. The CPUQuota setting is only available since systemd 213
+	calculateSystemdCPULimit := func() string {
+		if grp.CpuLimit.Count == 0 && grp.CpuLimit.Percentage == 0 {
+			return ""
+		}
+
+		cpuQuotaSnap := max(grp.CpuLimit.Count, 1) * grp.CpuLimit.Percentage
+		cpuQuotaMax := runtime.NumCPU() * 100
+		return fmt.Sprintf("CPUQuota=%d%%\n", min(cpuQuotaSnap, cpuQuotaMax))
+	}
+
+	// getAllowedCpusValue converts allowed CPUs array to a comma-delimited
+	// string that systemd expects it to be in. If the array is empty then
+	// an empty string is returned
+	getAllowedCpusValue := func() string {
+		if len(grp.CpuLimit.AllowedCPUs) == 0 {
+			return ""
+		}
+
+		allowedCpusValue := strings.Trim(strings.Join(strings.Fields(fmt.Sprint(grp.CpuLimit.AllowedCPUs)), ","), "[]")
+		return fmt.Sprintf("AllowedCPUs=%s\n", allowedCpusValue)
+	}
+
+	// AllowedCpus is only available for cgroupsv2
```

Comment:
also I don't see it in this PR because I assume a more logical place is in the actual servicestate implementation, but we should not allow initially configuring or later mutating/changing the value of AllowedCPUs if the current cgroups version does not support it. Unfortunately systems can switch between the unified (v2) and hybrid (v1) cgroups at will through the kernel command line so we can't assume that a system will always have the same cgroup version it used to.

Reply:
This is correct, the checks made for the correct cgroup version is done in cmd/snap as a part of the parameter handling. I don't know if we should handle a dynamically switched cgroup version, I'm thinking we just leave it and let systemd determine if it wants to use the ```AllowedCPUs``` parameter in the slice?

Reply:
I think unfortunately we probably still want to prevent changing values which are not enforced, similar to memory because this gives the wrong impression to the caller that the resource quota they are requesting will not be enforced and thus could have unintended side effects where the system is supposed to have some critical service constrained and now it is not constrained and ends up hanging the device instead of hitting OOM-killer for example

Reply:
Ok no problems, I'll keep that behaviour when I get to cmd/snap PR